sw/device/lib/dif/dif_spi_device.c

```diff
@@ -434,3 +480,539 @@ dif_result_t dif_spi_device_send(const dif_spi_device_t *spi,
+      spi->dev.base_addr, SPI_DEVICE_UPLOAD_STATUS_REG_OFFSET);
+  if (!bitfield_bit32_read(upload_status,
+                           SPI_DEVICE_UPLOAD_STATUS_CMDFIFO_NOTEMPTY_BIT)) {
+    return kDifUnavailable;
+  }
+  uint32_t cmd_item = mmio_region_read32(spi->dev.base_addr,
+                                         SPI_DEVICE_UPLOAD_CMDFIFO_REG_OFFSET);
+  *command =
+      bitfield_field32_read(cmd_item, SPI_DEVICE_UPLOAD_CMDFIFO_DATA_FIELD);
+  return kDifOk;
+}
+
+dif_result_t dif_spi_device_pop_flash_address_fifo(dif_spi_device_handle_t *spi,
+                                                   uint32_t *address) {
+  if (spi == NULL || address == NULL) {
+    return kDifBadArg;
+  }
+  uint32_t upload_status = mmio_region_read32(
+      spi->dev.base_addr, SPI_DEVICE_UPLOAD_STATUS_REG_OFFSET);
+  if (!bitfield_bit32_read(upload_status,
+                           SPI_DEVICE_UPLOAD_STATUS_ADDRFIFO_NOTEMPTY_BIT)) {
+    return kDifUnavailable;
+  }
+  *address = mmio_region_read32(spi->dev.base_addr,
+                                SPI_DEVICE_UPLOAD_ADDRFIFO_REG_OFFSET);
+  return kDifOk;
+}
+
+typedef struct dif_spi_device_flash_buffer_info {
+  size_t buffer_len;
+  ptrdiff_t buffer_offset;
+} dif_spi_device_flash_buffer_info_t;
+
+static dif_result_t dif_spi_device_get_flash_buffer_info(
+    dif_spi_device_flash_buffer_type_t buffer_type,
+    dif_spi_device_flash_buffer_info_t *info) {
+  switch (buffer_type) {
+    case kDifSpiDeviceFlashBufferTypeEFlash:
+      info->buffer_len = kDifSpiDeviceEFlashLen;
+      info->buffer_offset = kDifSpiDeviceEFlashOffset;
+      break;
+    case kDifSpiDeviceFlashBufferTypeMailbox:
+      info->buffer_len = kDifSpiDeviceMailboxLen;
+      info->buffer_offset = kDifSpiDeviceMailboxOffset;
+      break;
+    case kDifSpiDeviceFlashBufferTypeSfdp:
+      info->buffer_len = kDifSpiDeviceSfdpLen;
+      info->buffer_offset = kDifSpiDeviceSfdpOffset;
+      break;
+    case kDifSpiDeviceFlashBufferTypePayload:
+      info->buffer_len = kDifSpiDevicePayloadLen;
+      info->buffer_offset = kDifSpiDevicePayloadOffset;
+      break;
+    default:
+      return kDifBadArg;
+  }
+  return kDifOk;
+}
+
+dif_result_t dif_spi_device_read_flash_buffer(
+    dif_spi_device_handle_t *spi,
+    dif_spi_device_flash_buffer_type_t buffer_type, uint32_t offset,
+    size_t length, uint8_t *buf) {
+  if (spi == NULL || buf == NULL) {
+    return kDifBadArg;
+  }
+  dif_spi_device_flash_buffer_info_t info;
+  dif_result_t status =
+      dif_spi_device_get_flash_buffer_info(buffer_type, &info);
+  if (status != kDifOk) {
+    return status;
+  }
+  if (offset >= (info.buffer_offset + info.buffer_len) ||
+      length > (info.buffer_offset + info.buffer_len - offset)) {
+    return kDifBadArg;
+  }
+  ptrdiff_t offset_from_base =
+      SPI_DEVICE_BUFFER_REG_OFFSET + info.buffer_offset + offset;
+  mmio_region_memcpy_from_mmio32(spi->dev.base_addr, offset_from_base, buf,
```

Comment:
Is `spi_memcpy` still needed?

Reply:
`spi_memcpy` is still needed: It handles the FIFO pointers for generic mode.

`dif_spi_device_read_flash_buffer()` is purely for the RAM-style access used in flash and passthrough modes (i.e. it's not FIFO-style).